test/test_latest_time_policy.cpp

```diff
@@ -0,0 +1,237 @@
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+#include <gtest/gtest.h>
+
+#include <rclcpp/rclcpp.hpp>
+#include "message_filters/synchronizer.h"
+#include "message_filters/sync_policies/latest_time.h"
+
+using namespace message_filters;
+using namespace message_filters::sync_policies;
+
+struct Header
+{
+  rclcpp::Time stamp;
+};
+
+struct Msg
+{
+  Header header;
+  int data;
+};
+typedef std::shared_ptr<Msg> MsgPtr;
+typedef std::shared_ptr<Msg const> MsgConstPtr;
+
+namespace message_filters
+{
+namespace message_traits
+{
+template<>
+struct TimeStamp<Msg>
+{
+  static rclcpp::Time value(const Msg& m)
+  {
+    return m.header.stamp;
+  }
+};
+}
+}
+
+class Helper
+{
+public:
+  Helper()
+  : count_(0)
+  {}
+
+  void cb(const MsgConstPtr& p, const MsgConstPtr& q, const MsgConstPtr& r)
+  {
+    ASSERT_TRUE(p);
+    ASSERT_TRUE(q);
+    ASSERT_TRUE(r);
+
+    p_ = p; q_ = q; r_ = r;
+    ++count_;
+  }
+
+  MsgConstPtr p_, q_, r_;
+  uint16_t count_;
+};
+
+typedef LatestTime<Msg, Msg, Msg> Policy3;
+typedef Synchronizer<Policy3> Sync3;
+
+TEST(LatestTime, Leading)
+{
+  Sync3 sync(1);
+  Helper h;
+  sync.registerCallback(std::bind(&Helper::cb, &h, std::placeholders::_1,
+                                                   std::placeholders::_2,
+                                                   std::placeholders::_3));
+  std::vector<MsgPtr> p; p.reserve(8U);
+  std::vector<MsgPtr> q; q.reserve(4U);
+  std::vector<MsgPtr> r; r.reserve(2U);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    MsgPtr p_idx(std::make_shared<Msg>()); p_idx->data = idx; p.push_back(p_idx);
+    if(idx % 2U == 0U)
+    {
+      MsgPtr q_idx(std::make_shared<Msg>()); q_idx->data = idx; q.push_back(q_idx);
+    }
+    if(idx % 4U == 0U)
+    {
+      MsgPtr r_idx(std::make_shared<Msg>()); r_idx->data = idx; r.push_back(r_idx);
+    }
+  }
+
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    if(idx % 2U == 0U)
+    {
+      sync.add<1>(q[idx/2]);
+    }
+    if(idx % 4U == 0U)
+    {
+      sync.add<2>(r[idx/4]);
+    }
+    sync.add<0>(p[idx]);
+    
+    ASSERT_EQ(h.count_, idx);
```

Comment:
I see that other tests in this repository make heavy use of asserts, but I'd recommend using `EXPECT`s in places where a failure won't cause the test to crash. This way when the test does fail, it still continues and can have other failures too. That can be useful to find out the root cause of the failure, because there's more information.

Reply:
changed to `EXPECT_*`